tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -9720,6 +9720,69 @@ void hlsl::DiagnosePackingOffset(
   }
 }
 
+static bool IsIntegerTypeCSInputSemantic(const hlsl::SemanticDecl *sd) {
+  return
+    sd->SemanticName.equals("SV_DispatchThreadID") ||
+    sd->SemanticName.equals("SV_GroupIndex") ||
+    sd->SemanticName.equals("SV_GroupID") ||
+    sd->SemanticName.equals("SV_GroupThreadID");
+}
+
+static bool IsComputeShader(clang::Sema *self) {
+  StringRef shaderType(self->getLangOpts().HLSLProfile);
+  return shaderType.startswith_lower("cs_");
+}
+
+static bool
+HasValidCSInputSemanticType(clang::Sema *self, QualType &ty,
+                            const hlsl::SemanticDecl *semanticDecl) {
+  if (!IsIntegerTypeCSInputSemantic(semanticDecl))
+    return true;
+
+  QualType qty = ty.getCanonicalType();
+
+  const clang::Type *qtyPtr = qty.getTypePtrOrNull();
+  if (qtyPtr && qtyPtr->isScalarType() &&
+    qtyPtr->getScalarTypeKind() != Type::ScalarTypeKind::STK_Integral) {
+    return false;
+  }
+
+  if (hlsl::IsHLSLMatType(qty)) {
+    QualType matElemTy = GetHLSLMatElementType(qty);
+    const Type *matElemTyPtr = matElemTy.getTypePtrOrNull();
+    if (matElemTyPtr && matElemTyPtr->isScalarType() &&
+      matElemTyPtr->getScalarTypeKind() != Type::ScalarTypeKind::STK_Integral)
+      return false;
+  }
+
+  if (hlsl::IsHLSLVecType(qty)) {
+    QualType vecElemTy = GetHLSLVecElementType(qty);
+    const Type *vecElemtyPtr = vecElemTy.getTypePtrOrNull();
+    if (vecElemtyPtr && vecElemtyPtr->isScalarType() &&
+      vecElemtyPtr->getScalarTypeKind() != Type::ScalarTypeKind::STK_Integral)
+      return false;
+  }
+
+  return true;
+}
+
+void hlsl::DiagnoseSemanticType(clang::Sema *self, clang::QualType &type,
+                                hlsl::SemanticDecl *semanticDecl,
+                                bool isTypeDef, bool isLocalVar,
+                                const char *declarationType) {
+
+  if (isTypeDef || isLocalVar) {
+    self->Diag(semanticDecl->Loc, diag::err_hlsl_varmodifierna)
+        << "semantic" << declarationType;
```

Comment:
This is pre-existing of course, but I'm not sure it's consistent with our documentation as I read it here:
https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-variable-syntax#Semantic

which says that "The compiler ignores semantics unless they are declared on a global variable, or a parameter passed into a shader."